siwe.go

```diff
@@ -0,0 +1,462 @@
+	if m.requestID != nil {
+		ret := *m.requestID
+		return &ret
+	}
+	return nil
+}
+
+func (m *Message) GetResources() []string {
+	return m.resources
+}
+
+func GenerateNonce() string {
+	return uniuri.NewLen(16)
+}
+
+func isEmpty(str *string) bool {
+	return str == nil || len(strings.TrimSpace(*str)) == 0
+}
+
+const _SIWE_DOMAIN = "^(?P<domain>([^?#]*)) wants you to sign in with your Ethereum account:\\n"
+const _SIWE_ADDRESS = "(?P<address>0x[a-zA-Z0-9]{40})\\n\\n"
+const _SIWE_STATEMENT = "((?P<statement>[^\\n]+)\\n)?\\n"
+const _SIWE_URI = "(([^:?#]+):)?(([^?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))"
+
+var _SIWE_URI_LINE = fmt.Sprintf("URI: (?P<uri>%s?)\\n", _SIWE_URI)
+
+const _SIWE_VERSION = "Version: (?P<version>1)\\n"
+const _SIWE_CHAIN_ID = "Chain ID: (?P<chainId>[0-9]+)\\n"
+const _SIWE_NONCE = "Nonce: (?P<nonce>[a-zA-Z0-9]{8,})\\n"
+const _SIWE_DATETIME = "([0-9]+)-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[01])[Tt]([01][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9]|60)(\\.[0-9]+)?(([Zz])|([\\+|\\-]([01][0-9]|2[0-3]):[0-5][0-9]))"
+
+var _SIWE_ISSUED_AT = fmt.Sprintf("Issued At: (?P<issuedAt>%s)", _SIWE_DATETIME)
+var _SIWE_EXPIRATION_TIME = fmt.Sprintf("(\\nExpiration Time: (?P<expirationTime>%s))?", _SIWE_DATETIME)
+var _SIWE_NOT_BEFORE = fmt.Sprintf("(\\nNot Before: (?P<notBefore>%s))?", _SIWE_DATETIME)
+
+const _SIWE_REQUEST_ID = "(\\nRequest ID: (?P<requestId>[-._~!$&'()*+,;=:@%a-zA-Z0-9]*))?"
+
+var _SIWE_RESOURCES = fmt.Sprintf("(\\nResources:(?P<resources>(\\n- %s?)+))?$", _SIWE_URI)
+
+var _SIWE_MESSAGE = regexp.MustCompile(fmt.Sprintf("%s%s%s%s%s%s%s%s%s%s%s%s",
+	_SIWE_DOMAIN,
+	_SIWE_ADDRESS,
+	_SIWE_STATEMENT,
+	_SIWE_URI_LINE,
+	_SIWE_VERSION,
+	_SIWE_CHAIN_ID,
+	_SIWE_NONCE,
+	_SIWE_ISSUED_AT,
+	_SIWE_EXPIRATION_TIME,
+	_SIWE_NOT_BEFORE,
+	_SIWE_REQUEST_ID,
+	_SIWE_RESOURCES))
+
+func parseMessage(message string) (map[string]interface{}, error) {
+	match := _SIWE_MESSAGE.FindStringSubmatch(message)
+
+	if match == nil {
+		return nil, &ParsingFailed{"Message could not be parsed"}
+	}
+
+	result := make(map[string]interface{})
+	for i, name := range _SIWE_MESSAGE.SubexpNames() {
+		if i != 0 && name != "" && match[i] != "" {
+			result[name] = match[i]
+		}
+	}
+
+	originalAddress := result["address"].(string)
+	parsedAddress := common.HexToAddress(originalAddress)
+	if originalAddress != parsedAddress.String() {
+		return nil, &ParsingFailed{"Address must be in EIP-55 format"}
+	}
+
+	if val, ok := result["resources"]; ok {
+		result["resources"] = strings.Split(val.(string), "\n- ")[1:]
+	}
+
+	return result, nil
+}
+
+func ParseMessage(message string) (*Message, error) {
+	result, err := parseMessage(message)
+	if err != nil {
+		return nil, err
+	}
+
+	parsed, err := InitMessage(
+		result["domain"].(string),
+		result["address"].(string),
+		result["uri"].(string),
+		result["version"].(string),
+		result,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return parsed, nil
+}
+
+func signHash(data []byte) common.Hash {
```

Comment:
Should be a method of `Message`, and probably call it `eip191Hash` like in `siwe-rs`